acados/ocp_nlp/ocp_nlp_globalization_funnel.c

```diff
@@ -0,0 +1,637 @@
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.;
+ */
+
+
+#include "acados/ocp_nlp/ocp_nlp_globalization_common.h"
+#include "acados/ocp_nlp/ocp_nlp_globalization_funnel.h"
+#include "acados/ocp_nlp/ocp_nlp_common.h"
+#include "acados/ocp_nlp/ocp_nlp_sqp.h"
+
+// external
+#include <assert.h>
+#include <math.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#if defined(ACADOS_WITH_OPENMP)
+#include <omp.h>
+#endif
+
+// blasfeo
+#include "blasfeo/include/blasfeo_d_aux.h"
+#include "blasfeo/include/blasfeo_d_blas.h"
+// acados
+#include "acados/utils/mem.h"
+
+
+/************************************************
+ * options
+ ************************************************/
+
+acados_size_t ocp_nlp_globalization_funnel_opts_calculate_size(void *config_, void *dims_)
+{
+    ocp_nlp_dims *dims = dims_;
+    ocp_nlp_config *config = config_;
+
+    acados_size_t size = 0;
+
+    size += sizeof(ocp_nlp_globalization_funnel_opts);
+
+    size += ocp_nlp_globalization_opts_calculate_size(config, dims);
+
+    return size;
+}
+
+void ocp_nlp_globalization_funnel_opts_initialize_default(void *config_, void *dims_, void *opts_)
+{
+    ocp_nlp_dims *dims = dims_;
+    ocp_nlp_globalization_funnel_opts *opts = opts_;
+    ocp_nlp_globalization_opts *globalization_opts = opts->globalization_opts;
+    ocp_nlp_globalization_config *config = config_;
+
+    ocp_nlp_globalization_opts_initialize_default(config, dims, globalization_opts);
+
+    // funnel method opts
+    opts->initialization_increase_factor = 15.0;
+    opts->initialization_upper_bound = 1.0;
+    opts->sufficient_decrease_factor = 0.9;
+    opts->kappa = 0.9;
+    opts->fraction_switching_condition = 1e-3;
+    opts->initial_penalty_parameter = 1.0;
+    opts->penalty_contraction = 0.5;
+    opts->penalty_eta = 1e-6;
+    opts->type_switching_condition = false; // use ipopt/gould type of switching
+
+    
+    return;
+}
+
+
+void ocp_nlp_globalization_funnel_opts_set(void *config_, void *opts_, const char *field, void* value)
+{    
+    ocp_nlp_globalization_funnel_opts *opts = opts_;
+    ocp_nlp_globalization_config *config = config_;
+
+    if (!strcmp(field, "funnel_initialization_increase_factor"))
+    {
+        double* funnel_initialization_increase_factor = (double *) value;
+        if (*funnel_initialization_increase_factor <= 1.0)
+        {
+            printf("\nerror: ocp_nlp_sqp_opts_set: invalid value for funnel_initialization_increase_factor field, need double > 1, got %f.", *funnel_initialization_increase_factor);
```

Comment:
message should contain correct funciton name.

Reply:
I think you fixed that.